### ISSUE TITLE
Reform how Leo handles git time stamps

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#1126]
Inspired by [this post](https://groups.google.com/d/msg/leo-editor/K6lTBuap4o4/z1B_Ufl4BAAJ) by Vitalije.

**Status**

leoVersion.py now contains only two static module-level vars: version and static_date.

app.compute_signon calls g.getGitVersion to do all the work.  This function does not assume that git log entries are in a fixed order.

The 5.9 code base must remain strictly unchanged. I'll remove /commit_timestamp.json and the leo/extensions/hooks folder *after* Leo 5.9 goes out the door. I'll close this issue after doing so.

No changes should be needed for setup.py.  The script that makes daily builds may end up using static_date.  That script could, if desired, hack the leo.core.leoVersion.date/build vars.

**Comment by maphew** [comment]
Related is the version number used in setup.py for PyPi:

    #{tag} {distance-from-tag} {current commit hash}
    5.8.1b1   dev1342   e6c414ca

In usage (edited for brevity): 
```
> git checkout devel

> python setup.py check
...Normalizing '5.8.1-b1-dev1342' to '5.8.1b1.dev1342'

> leo-m --version
Leo 5.9-final-devel, build 20190427172213, Sat Apr 27 17:22:13 CDT 2019
```

Now same for master:
```
> git checkout master

> python setup.py check
... Normalizing '5.9.b2' to '5.9b2'

> leo-m --version
Leo 5.9-b2, build 20190409061733, Tue Apr  9 06:17:32 UTC 2019
```

_Commit_ is dropped in setup.py's version report because I wanted to be less noisy, not give users too much to remember. There's no technical reason I'm aware of that it couldn't be added.

Setup's functions are defined in `get_version()` which cascades internally to `git_version()` and externally to `g.gitDescribe()`.

Bottom line, we should use the same version scheme. I'll follow your lead. (I didn't before because I couldn't follow where Leo was pulling it's version from. I understand more now.)  I'm agnostic on whether it's build date or commit hash that follows the _Major.Minor.X_.

**Comment by edreamleo** [comment]
@maphew Thanks for your comments, Matt.

**Comment by vitalije** [comment]
@maphew 
Am I correct to assume that building PyPy Leo package is never done without .git folder being present?
In other words, setup.py can rely on the presence of .git folder, is this true?

If that is the case, then the simplest thing would be to just delete commit_timestamp.json file from git and leave hooks as they are, and in setup.py we can regenerate this commit_timestamp.json and add it to the package.

**Comment by edreamleo** [comment]
@maphew @vitalije Imo, the simplest thing would be to delete commit_timestamp.json and leo/extensions/hooks.  Any comments?

**Comment by tbnorth** [comment]
On Sun, 28 Apr 2019 15:23:49 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> @maphew @vitalije Imo, the simplest thing would be to delete
> commit_timestamp.json and leo/extensions/hooks.  Any comments?

I think the goal of the timestamp was to allow someone to report a
problem, a dev. pushes a quick fix, the reporter downloads
https://github.com/leo-editor/leo-editor/archive/devel.zip, and reports
"yes it's fixed / no it's still broken in 201904041234" (unambiguous),
as opposed to "it's fixed / broken in the one I downloaded 5 minutes
ago, when did you upload the fix exactly?"...

But if the mechanism's too expensive or your ok with only offering
quick fixes via git pulls, it would certainly be simpler to do without
it.

Cheers -Terry

**Comment by edreamleo** [comment]
There was a long discussion of this [in this thread](https://groups.google.com/d/msg/leo-editor/o1fKiy-rymA/7JWzcvHpBAAJ).

**Terry**: The point of it was to give useful, *fine grained* version ID in non-git contexts.

I'm not sure whether this includes daily builds.

**Comment by maphew** [comment]
@vitalije Unfortunately setup.py can't rely on .git folder. It's executed/processed by pip when installing. The behind the scenes process on the target machine is something like: "export to source archive package >> extract to temp folder >> build package in there, creating a dist >> copy the dist to the right places for current architecture.". Using wheels sidesteps some of the middle things, but there's still no .git folder present.

We could put git commit info in the version tag/package name and thereby get the fine grained version info on hand for pip (will be reported by `pip list`), but that doesn't help the Leo log window.

There is https://github.com/pypa/setuptools_scm which I've read through a few times and made some initial experiments last year. If I read the docs correctly it means using it inside Leo and not just for packaging, which is what made me put the experiments away. Maybe it's not as formidable as it looks, just a lot of words, and the implementation is actually easy.

**Comment by vitalije** [comment]
@maphew Sorry, but I didn't ask the right question. I am not familiar with the way you create PyPy package. I thought that setup.py is invoked to do that. What I wanted to know is whether at the time of creating of PyPy package, script that does this packaging (whatever it may be called), can rely on the .git folder? If it can, then in that script we can create commit_timestamp.json from the git information and add it to the package. In that case Leo for the purpose of reporting timestamp in Log pane, can try to access git information (if it is installed from github.com) and if that fails to report from commit_timestamp.json.

Anyway if the reason for having time reported in Log was to allow user to report finer version details, it really makes sense only if the user installed Leo through git. If the installation is from PyPy then the timestamp is irrelevant because pypy version tells us all we need to know. Leo can just report its PyPy version in Log.

@edreamleo is it acceptable that Leo should report timestamp only if it is installed through git?

**Comment by edreamleo** [comment]
@vitalije Yes.  Reporting a timestamp could be done only when git is present.  I'll say more in leo-editor.

**Comment by maphew** [comment]
This breaks Leo when run from source and no .git folder present. The export below corresponds to commit 802f42eb170b8555ec6224bcfbf86d0d98fbd94f (HEAD -> py3, origin/py3).

Move `.git` somewhere else then:

```
$ python launchLeo.py

fatal: not a git repository (or any of the parent directories): .git
Traceback (most recent call last):
  File "launchLeo.py", line 8, in <module>
    leo.core.runLeo.run()
  File "D:\x\dist\leo-5.8.1b1.dev1634\leo\core\runLeo.py", line 72, in run
    g.app.loadManager.load(fileName, pymacs)
  File "D:\x\dist\leo-5.8.1b1.dev1634\leo\core\leoApp.py", line 2251, in load
    lm.doPrePluginsInit(fileName, pymacs)
  File "D:\x\dist\leo-5.8.1b1.dev1634\leo\core\leoApp.py", line 2433, in doPrePluginsInit
    g.app.computeSignon() # Set app.signon/signon1 for commanders.
  File "D:\x\dist\leo-5.8.1b1.dev1634\leo\core\leoApp.py", line 968, in computeSignon
    author, commit, date = g.getGitVersion()
  File "D:\x\dist\leo-5.8.1b1.dev1634\leo\core\leoGlobals.py", line 4980, in getGitVersion
    cwd=g.app.loadDir, shell=True)
  File "c:\tools\miniconda3\envs\leo-dev\lib\subprocess.py", line 389, in check_output
    **kwargs).stdout
  File "c:\tools\miniconda3\envs\leo-dev\lib\subprocess.py", line 481, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'git log -n 1 --date=iso' returned non-zero exit status 128.
```

**Comment by edreamleo** [comment]
@maphew Good catch.  Should be fixed at c78f73 in py3.

**Comment by edreamleo** [comment]
This issue can be closed because the code has been merged into devel.
